src/main/flight/autopilot.c

```diff
@@ -22,38 +22,141 @@
 
 #include "platform.h"
 #include "build/debug.h"
+#include "common/axis.h"
+#include "common/filter.h"
 #include "common/maths.h"
+#include "common/vector.h"
+#include "fc/rc.h"
+#include "fc/runtime_config.h"
 
 #include "flight/imu.h"
-#include "flight/pid.h"
 #include "flight/position.h"
 #include "rx/rx.h"
+#include "sensors/gyro.h"
 
 #include "autopilot.h"
 
 #define ALTITUDE_P_SCALE  0.01f
 #define ALTITUDE_I_SCALE  0.003f
 #define ALTITUDE_D_SCALE  0.01f
 #define ALTITUDE_F_SCALE  0.01f
+#define POSITION_P_SCALE  0.0012f
+#define POSITION_I_SCALE  0.0001f
+#define POSITION_D_SCALE  0.0015f
+#define POSITION_A_SCALE  0.0008f
+#define UPSAMPLING_CUTOFF_HZ 5.0f
 
 static pidCoefficient_t altitudePidCoeffs;
+static pidCoefficient_t positionPidCoeffs;
+
 static float altitudeI = 0.0f;
 static float throttleOut = 0.0f;
 
+typedef struct {
+    bool isStopping;
+    float distance;
+    float previousDistance;
+    float previousVelocity;
+    float integral;
+    float pidSum;
+    pt1Filter_t velocityLpf;
+    pt1Filter_t accelerationLpf;
+} earthFrame_t;
+
+typedef enum {
+    EW = 0,
+    NS
+} axisEF_t;
```

Comment:
Usually "North" is the first component, followed by "East" (and then "Up" or "Down") - NEU or NED. Can you still change this or is it too risky to make mistakes in the process?

Reply:
Problem is that BF uses NEU, which is nonstandard. It is probably easier to change it to ENU than NED ...